### PR TITLE
MAJOR CHANGE: Adding directionality to the edge...

### PR DESCRIPTION
**Pull request by sirpoovey**
Edges will now hold a direction.  This means that we will now store the edge both in the source and destination vertex edge lists.  This allows for traversal over in edges to be much easier which is crucial for performance of many algorithms (such as link prediction).

There was an old inherent assumption throughout STINGER that the graph is undirected.  In some cases this PR will address this (e.g. PageRank and egonets now account for the directionality appropriately).  However in other cases (such as pagerank_updating or betweenness centrality) the code is updated to maintain the same functionality as before.  My proposal is to address these issues as we encounter them, as it is unclear which algorithms and RPC calls were built to be only for undirected graphs.

This change is BREAKING to the API.  Several old and unused functions have been moved to stinger_deprecated.  The entire standalone directory was removed as it is was part of the STINGER research project and not the STINGER client/server graph database this project is intended to support.  One of the biggest changes for anyone using the standalone approach is that set_initial_edges now expects a graph with directionality to be passed in.  That means that the CSR format should also have edges stored in both the source and destination vertex edge list.  Creating a STINGER without the directionality properly set may cause some algorithms to give incorrect results.  If in doubt, at least mark out edges by ORing the neighbor field with STINGER_EDGE_DIRECTION_OUT.

All changes have been tested using the updated test suites.  Also a server/client version has been run with a sample graph.  This is ready for review.

### CODE REVIEW

**Comment by sirpoovey** [comment]
@davidediger @robmccoll @tgoodyear - This code is ready for review and comment

File: lib/stinger_core/src/stinger.c [hunk 1]
```diff
@@ -276,15 +276,13 @@ stinger_max_active_vertex(const struct stinger * S) {
   uint64_t out = 0;
   OMP("omp parallel") {
     uint64_t local_max = 0;
-    OMP("omp for")
```
Comment (ehein6) [inline lib/stinger_core/src/stinger.c]: Why was this removed? I think it can stay.

File: lib/stinger_core/src/stinger.c [hunk 2]
```diff
@@ -972,85 +992,106 @@ etype_begin (stinger_t * S, int64_t v, int etype)
...
-	e->weight = weight;
-	break;
-      case EDGE_WEIGHT_INCR:
-	stinger_int64_fetch_add(&(e->weight), weight);
-	break;
+    int64_t weight = readfe (&(e->weight));
+    
+    if (direction & STINGER_EDGE_DIRECTION_OUT) {
+      if (operation & EDGE_WEIGHT_SET) {
+        //e->weight = weight;
```
Comment (ehein6) [inline lib/stinger_core/src/stinger.c]: Why is the update of e->weight deferred until later? Anyways, commented out code should not be checked in.

File: lib/stinger_core/src/stinger.c [hunk 3]
```diff
@@ -972,85 +992,106 @@ etype_begin (stinger_t * S, int64_t v, int etype)
...
+        if (ts > eb->largeStamp)
+          eb->largeStamp = ts;
+        writeef(&eb->smallStamp, smallStamp);
+      }
 
-    e->timeRecent = ts;
-
+      e->timeRecent = ts;
+    }
+    writeef((uint64_t *)&(e->weight), (uint64_t)weight);
```
Comment (ehein6) [inline lib/stinger_core/src/stinger.c]: Need a comment here explaining why write to e->weight was deferred from above.

File: lib/stinger_core/src/stinger.c [hunk 4]
```diff
@@ -972,85 +992,106 @@ etype_begin (stinger_t * S, int64_t v, int etype)
...
-    e->neighbor = neighbor;
-  }
+    if (direction & STINGER_EDGE_DIRECTION_OUT) {
+      e->neighbor = e->neighbor & ~STINGER_EDGE_DIRECTION_OUT;
+      stinger_outdegree_increment_atomic(S, eb->vertexID, -1);
+    } else if (direction & STINGER_EDGE_DIRECTION_IN) {
+      e->neighbor = e->neighbor & ~STINGER_EDGE_DIRECTION_IN;
+      stinger_indegree_increment_atomic(S, eb->vertexID, -1);
+    }
+    // TODO: Make sure there is no race condition here and that e->neighbor will be up to date.
```
Comment (ehein6) [inline lib/stinger_core/src/stinger.c]: Is there a race condition here or not?

File: lib/stinger_core/src/stinger.c [hunk 5]
```diff
@@ -1410,6 +1465,8 @@ stinger_remove_edge_pair (struct stinger *G,
 
 MTA("mta parallel default")
 
+/* FIXME: Doesn't follow the STINGER direction standards.  Only inserts out edges */
```
Comment (ehein6) [inline lib/stinger_core/src/stinger.c]: Shouldn't this be fixed to also insert the in edges?

File: lib/stinger_core/src/stinger.c [hunk 6]
```diff
@@ -1466,7 +1525,7 @@ stinger_set_initial_edges (struct stinger *G,
   nblk_total = blkoff[nv];
 
   block = xcalloc (nblk_total, sizeof (*block));
-  OMP ("omp parallel for")
+  /*OMP ("omp parallel for")
```
Comment (ehein6) [inline lib/stinger_core/src/stinger.c]: Remove commented code

File: lib/stinger_core/src/stinger.c [hunk 7]
```diff
@@ -2119,11 +2259,16 @@ int64_t
...
     stinger_remove_edge_pair(G,STINGER_EDGE_TYPE,STINGER_EDGE_SOURCE,STINGER_EDGE_DEST);
-  } STINGER_PARALLEL_FORALL_EDGES_OF_VTX_END();
+  } STINGER_PARALLEL_FORALL_OUT_EDGES_OF_VTX_END();
+
+  STINGER_PARALLEL_FORALL_IN_EDGES_OF_VTX_BEGIN(G,vtx_id) {
+    stinger_remove_edge_pair(G,STINGER_EDGE_TYPE,STINGER_EDGE_DEST,STINGER_EDGE_SOURCE);
+  } STINGER_PARALLEL_FORALL_IN_EDGES_OF_VTX_END();
 
   // Remove remaining in edges
+  /* JP: This loop should no longer be necessary now that in edges will be capture in above loop
```
Comment (ehein6) [inline lib/stinger_core/src/stinger.c]: Agreed, remove the loop.

File: lib/stinger_core/src/stinger.c (hunk 2 above)
Comment (sirpoovey) [inline lib/stinger_core/src/stinger.c]: edge weight is used as a lock on the edge data.  This is keep consistency with deletion of edges.

File: lib/stinger_core/src/stinger.c (hunk 5 above)
Comment (sirpoovey) [inline lib/stinger_core/src/stinger.c]: This will put in the in edges if they are provided.  This comment can be removed since we got rid of the standalone directory

**Comment by ehein6** [comment]
Finished my review, however I'll admit I couldn't really tell what was going on in stinger_remove_edge, perhaps someone can walk me through it.

**Comment by davidediger** [comment]
Thanks @sirpoovey .  You're the best.
